### PR TITLE
Multigenerational touch values, Remake of the dynamic threshold algo

### PR DESCRIPTION
**Pull request by marcos-diaz**
Also the tune values are not "threshold" values anymore, but "sensitivity" values (from less sensitive to most sensitive) for consistency with the other tune settings

### CODE REVIEW

File: src/touch.c [hunk 1]
```diff
@@ -35,72 +55,68 @@ void touch_init() {
...
+bool touch_status() {
+    uint32_t elapsed = touch_get_elapsed();
+    // Determine threshold.
+    float threshold = (
+        threshold_from_config > 0 ?
+        threshold_from_config :
+        touch_get_dynamic_threshold(elapsed)
+    );
+    // In case of timeout.
+    if (elapsed == 0) elapsed = threshold / CFG_TOUCH_DYNAMIC_PEAK_RATIO;
```
Comment (marcos-diaz) [inline src/touch.c]: This could be improved.

File: src/headers/touch.h [hunk 2]
```diff
@@ -3,17 +3,29 @@
...
-#define CFG_TOUCH_DYNAMIC_MIN 3 // Microseconds
-#define CFG_TOUCH_DYNAMIC_FACTOR 2
-
-#define CFG_TOUCH_THRESHOLD_0 0  // Automatic.
-#define CFG_TOUCH_THRESHOLD_1 2  // Microseconds.
-#define CFG_TOUCH_THRESHOLD_2 3  // Microseconds.
-#define CFG_TOUCH_THRESHOLD_3 5  // Microseconds.
-#define CFG_TOUCH_THRESHOLD_4 8  // Microseconds.
+
+#define CFG_GEN0_TOUCH_TIMEOUT 20  // Microseconds.
```
Comment (przymi) [inline src/headers/touch.h]: Please provide either documentation or link to documentation as to which PCB versions belong to GEN0 and to GEN1

File: src/headers/touch.h [hunk 3]
```diff
@@ -3,17 +3,29 @@
...
+#define CFG_GEN0_TOUCH_SENS_3 3   // Microseconds. High.
+#define CFG_GEN0_TOUCH_SENS_4 2   // Microseconds. Very high.
+
+#define CFG_GEN1_TOUCH_SENS_0 0   // Automatic.
+#define CFG_GEN1_TOUCH_SENS_1 40  // Microseconds. Low.
+#define CFG_GEN1_TOUCH_SENS_2 25  // Microseconds. Mid.
+#define CFG_GEN1_TOUCH_SENS_3 15  // Microseconds. High.
+#define CFG_GEN1_TOUCH_SENS_4 10  // Microseconds. Very high.
+
+#define CFG_GEN0_TOUCH_DYNAMIC_MIN 3  // Microseconds
```
Comment (przymi) [inline src/headers/touch.h:24]: What were these for? Please comment on the impact of these constants

File: src/touch.c [hunk 4]
```diff
@@ -2,27 +2,47 @@
...
 #include <pico/stdlib.h>
 #include "config.h"
 #include "touch.h"
 #include "pin.h"
 #include "helper.h"
 
 uint8_t loglevel = 0;
-uint8_t threshold_config = 0;
-uint8_t peak = 0;
+uint8_t threshold_from_config = 0;
```
Comment (przymi) [inline src/touch.c]: This should now be sensitivity_config

File: src/touch.c [hunk 5]
```diff
@@ -35,72 +55,70 @@ void touch_init() {
...
+    } else {
+        elapsed = threshold + 1;  // Using threshold from previous cycle.
+    }
     // Debug.
     if (loglevel >= 2) {
-        static uint16_t x= 0;
+        static uint16_t x = 0;
         x++;
-        if (!(x % 20)) printf("%i %i\n", timing, peak);
+        if (!(x % 40)) printf("%i %.2f\n", elapsed, threshold);
```
Comment (przymi) [inline src/touch.c]: What is this magic number 40?

**Review by przymi — CHANGES_REQUESTED** [review]

**Pull request merged**